osquery/TPipe.py

```diff
@@ -0,0 +1,225 @@
+"""This source code is licensed under the BSD-style license found in the
+LICENSE file in the root directory of this source tree. An additional grant
+of patent rights can be found in the PATENTS file in the same directory.
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import win32event
+import win32pipe
+import win32file
+import win32api
+import winerror
+import pywintypes
+
+from thrift.transport.TTransport import TTransportBase
+from thrift.transport.TTransport import TTransportException
+from thrift.transport.TTransport import TServerTransportBase
+
+
+class TPipeBase(TTransportBase):
+    """Base class for Windows TPipe transport"""
+
+    def __init__(self):
+        self._handle = None
+
+    def close(self):
+        """
+        Generic close method, as both server and client rely on closing pipes
+        in the same way
+        """
+        if self._handle is not None:
+            win32pipe.DisconnectNamedPipe(self._handle)
+            self._handle = None
+
+
+class TPipe(TPipeBase):
+    """Client for communicating over Named Pipes"""
+
+    def __init__(self, pipe_name, timeout=5, max_attempts=5):
+        """
+        Initialize a TPipe client
+
+        @param pipe_name(string)  The path of the pipe to connect to.
+        @param timeout(int)  The time to wait for a named pipe connection.
+        @param max_attempts(int)  Maximum number of connection attempts.
+        """
+        self._handle = None
+        self._pipe_name = pipe_name
+        self._timeout = timeout
+        self._max_conn_attempts = max_attempts
+
+    def set_handle(self, h):
+        self._handle = h
+
+    def is_open(self):
+        return self._handle is not None
+
+    def open(self):
+        if self.is_open():
+            raise TTransportException(TTransportException.ALREADY_OPEN)
+
+        h = None
+        conns = 0
+        while conns < self._max_conn_attempts:
+            try:
+                h = win32file.CreateFile(
+                        self._pipe_name,
+                        win32file.GENERIC_READ | win32file.GENERIC_WRITE,
+                        0,
+                        None,
+                        win32file.OPEN_EXISTING,
+                        win32file.FILE_FLAG_OVERLAPPED,
+                        None)
+            except Exception as e:
+                if e[0] != winerror.ERROR_PIPE_BUSY:
```

Comment:
This line gives an IndexError - I think it should just be 'if e 1= ..'. 

```
Traceback (most recent call last):
  File "C:\Python36-32\lib\site-packages\osquery-3.0.1-py3.6.egg\osquery\TPipe.py", line 76, in open
pywintypes.error: (2, 'CreateFile', 'The system cannot find the file specified.')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\ProgramData\osquery\extensions\bb_e2e_test.ext", line 69, in <module>
    osquery.start_extension(name="bb_e2e_tests", version="1.0.0")
  File "C:\Python36-32\lib\site-packages\osquery-3.0.1-py3.6.egg\osquery\management.py", line 198, in start_extension
  File "C:\Python36-32\lib\site-packages\osquery-3.0.1-py3.6.egg\osquery\extension_client.py", line 62, in open
  File "C:\Python36-32\lib\site-packages\thrift\transport\TTransport.py", line 153, in open
    return self.__trans.open()
  File "C:\Python36-32\lib\site-packages\osquery-3.0.1-py3.6.egg\osquery\TPipe.py", line 78, in open
TypeError: 'error' object does not support indexing
```

Reply:
Ok will take a look. Just as a heads up though, looks like you’re testing this with Python 3,m given The path to your binary, and I’m honestly not sure how this code will behave. That index does look correct to me though, as e[0] will be the error code which is what we are intending to check against, but I’ll see if I can sort out what’s up with the exception. Thanks for testing!

Reply:
Ah, good catch. In fact, awesome to know this works out of the box in py3 :).